---
title: Installing Packages
author: Jeffrey Brabec
date: '2020-08-26'
slug: installing-packages
categories:
  - Tutorials
tags: []
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```


## Installing Packages

As you start doing more advanced analyses in R (e.g. gene expression or imaging analysis; machine learning), there are times where you'll need to install new packages that were written specifically to do those analyses. These packages can add completely new tools to your R toolbox, or they will help optimize functions or outputs from base R and other R packages. 

So how do we install new packages? The most common way to download packages is done using this syntax:
```{r eval=FALSE}
install.packages("tidyverse")
```

This function, `install.packages()`, searches "The Comprehensive R Archive Network" or [CRAN](https://cran.r-project.org) for the package(s) you pass to it. CRAN is a file transfer protocol [(FTP)](https://en.wikipedia.org/wiki/File_Transfer_Protocol) server with updated versions of all R packages maintained on CRAN.

There will be times where you start using another lab's workflow and you might find that they use multiple packages that you have yet to install. Instead of typing `install.packages()` over and over again, you can provide a vector of the package names you'd like to install, and the function will iteratively install the packages listed:
```{r eval=FALSE}
to_install <- c("tidyverse", "gprofiler2", "cowplot")
install.packages(to_install)
#OR
install.packages(c("tidyverse", "gprofiler2", "cowplot"))
```

The same thing can be done when you want to load your packages at the beginning of your workflow. Sometimes you could have upwards of 15 packages that you're working with and it's inefficient to individually load each one:
```{r eval=FALSE}
to_load <- c("tidyverse", "gprofiler2", "cowplot", "Matrix", "tidymodels")
library(to_load)
#OR
library(c("tidyverse", "gprofiler2", "cowplot", "Matrix", "tidymodels"))
```

# Devtools
Sometimes you'll run across someone who is in the process of developing a new package which means it hasn't been released on CRAN yet. They'll often make development versions of their code available on GitHub for people to install and use prior to release of the full version. This practice ensures that a useful tool is accessible as soon as possible, and it also allows people to road test the code and find any bugs prior to full release.

To install a package that is still in development from GitHub you'll need to first install the `remotes` pacakge from CRAN. It has several handy tools for downloading and updating packages from various repositories!

In this case I want to download the TidyTuesdayR package which lets me easily read each week's [TidyTuesday](https://github.com/rfordatascience/tidytuesday) data directly into R.
```{r eval=FALSE}
install.packages("remotes")
install_github("thebioengineer/tidytuesdayR") 
```
To install a package from GitHub you need to provide the username of the person developing the package and the repository in which they have the package. In this case "thebioengineer" is the one developing this package and he has all the relevant code needed to compile the package in the repository "tidytuesdayR". 

However, as of June 26th the tidytuesdayR package is on CRAN so it's now available to install using `install.packages()`!

## Bioconductor

As biological scientists there is one final package repository that we'll want to use if we're doing any kind of genetic or genomic analyses and that is [Bioconductor](http://bioconductor.org). Bioconductor specifically maintains packages that help with biological analyses. Packages kept on Biocondoucter will help you do everythign from downloading data from online databases to doing advanced statistics on biological data. As we continue diving into R we'll do a deep dive on some of the most useful packages they provide. 

To install Bioconductor:
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE)) 
    install.packages("BiocManager")
BiocManager::install(version = "3.11")
```

This snippet of code first checks if the package "BiocManager" is already downloaded. If it isn't than the package is installed. Once its installed, the package has its own "install" function that allows you to update it to the latest version which, at the time of writing, was 3.11.

That last line of code in the previous section is the exact syntax we'll use to download packages from Bioconductor:
```{r eval=FALSE}
BiocManager::install("GEOquery")
```

`GEOquery` is a useful package that we'll use in future sessions. It provides an interface to the Gene Expression Omnibus or [GEO](https://www.ncbi.nlm.nih.gov/geo/) which is a repository of gene expression data from NIH funded projects.

Those are the main ways that we'll install packages in R! I hope you found this helpful!